File: leetcode/word_break/word_break.cpp
/*
 * Time Complexity: O(N^3)
 *   - N is the length of the string s.
 *   - For each position, we try every word and substring comparison.
 *   - Memoization reduces redundant calls, but worst case is cubic due to substring and set operations.
 *
 * Space Complexity: O(N)
 *   - For the positionVisited set and recursion stack.
 */

#include <string>
#include <vector>
#include <unordered_set>

using std::string;
using std::vector;
using std::unordered_set;

class Solution {
private:
    bool fn(const string& s, const vector<string>& wordDict, unordered_set<int>& positionVisited, int pos) {
        if(pos == s.size()) return true;
        positionVisited.insert(pos);

        int endPos;
        for(string word : wordDict) {
            endPos = word.size() + pos;
            if(
                endPos <= s.size() &&
                positionVisited.find(endPos) == positionVisited.end() &&
                word == s.substr(pos,endPos-pos) &&
                fn(s, wordDict, positionVisited, endPos)
            ) return true;
        }
        return false;
    }

public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<int> positionVisited;
        return fn(s, wordDict, positionVisited, 0);
    }
};

File: leetcode/word_break/word_break.java
package word_break;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    private Set<Integer> positionVisited;
    private String s;
    private List<String> wordDict;

    private boolean fn(int pos) {
        if(pos == s.length()) return true;
        positionVisited.add(pos);

        for(String word : wordDict) {
            int endPos = word.length() + pos;
            if(
                endPos <= s.length() &&
                !positionVisited.contains(endPos) &&
                word.equals(s.substring(pos, endPos)) &&
                fn(endPos)
            ) return true;
        }
        return false;
    }
    public boolean wordBreak(String s, List<String> wordDict) {
        this.positionVisited = new HashSet<>();
        this.s = s;
        this.wordDict = wordDict;
        return fn(0);
    }
}
